File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> idxMap = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            idxMap.put(players[i], i);
        }
        
        for (String calling : callings) {
            int idx = idxMap.get(calling);
            String tmp = players[idx - 1];
            players[idx - 1] = players[idx];
            players[idx] = tmp;
            
            idxMap.put(players[idx], idx);
            idxMap.put(players[idx - 1], idx - 1);
        }
        
        return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.kt
class Solution {
    fun solution(players: Array<String>, callings: Array<String>): Array<String> {
        val idxMap = mutableMapOf<String,Int>()
        for (i in players.indices) {
            idxMap[players[i]] = i
        }
        
        for (calling in callings) {
            val idx = idxMap[calling]!!
            val tmp = players[idx - 1]
            players[idx - 1] = players[idx]
            players[idx] = tmp
            
            idxMap[players[idx]] = idx
            idxMap[players[idx - 1]] = idx - 1
        }
        
        return players
    }
}